solutionbox/code_free_ml/analyze_data.py

```diff
@@ -0,0 +1,823 @@
+              max(numerical_results[col_name]['max'],
+                  float(parsed_line[col_name])))
+            numerical_results[col_name]['count'] += 1
+            numerical_results[col_name]['sum'] += float(parsed_line[col_name])
+          elif transform == KEY_TRANSFORM:
+            pass
+          else:
+            raise ValueError('Unknown transform %s' % transform)
+
+  # Write the vocab files. Each label is on its own line.
+  vocab_sizes = {}
+  for name, label_count in six.iteritems(vocabs):
+    # Labels is now the string:
+    # label1,count
+    # label2,count
+    # ...
+    # where label1 is the most frequent label, and label2 is the 2nd most, etc.
+    labels = '\n'.join(['%s,%d' % (label, count)
+                        for label, count in sorted(six.iteritems(label_count),
+                                                   key=lambda x: x[1],
+                                                   reverse=True)])
+    file_io.write_string_to_file(
+        os.path.join(args.output_dir, VOCAB_ANALYSIS_FILE % name),
+        labels)
+
+    vocab_sizes[name] = {'vocab_size': len(label_count)}
+
+  # Update numerical_results to just have min/min/mean
+  for col_name in numerical_results:
+    mean = (numerical_results[col_name]['sum'] /
+            float(numerical_results[col_name]['count']))
+    del numerical_results[col_name]['sum']
+    del numerical_results[col_name]['count']
+    numerical_results[col_name]['mean'] = mean
+
+  # Write the stats file.
+  numerical_results.update(vocab_sizes)
+  stats = {'column_stats': numerical_results, 'num_examples': num_examples}
+  file_io.write_string_to_file(
+      os.path.join(args.output_dir, STATS_FILE),
+      json.dumps(stats, indent=2, separators=(',', ': ')))
+
+
+def check_schema_transform_match(schema, features):
+  """Checks that the transform and schema do not conflict.
+
+  Args:
+    schema: schema file
+    features: features file
+
+  Raises:
+    ValueError if transform cannot be applied given schema type.
+  """
+  num_key_transforms = 0
+  num_target_transforms = 0
+
+  for col_schema in schema:
+    col_name = col_schema['name']
+    col_type = col_schema['type'].lower()
+
+    transform = features[col_name]['transform']
+    if transform == KEY_TRANSFORM:
+      num_key_transforms += 1
+      continue
+    elif transform == TARGET_TRANSFORM:
+      num_target_transforms += 1
+      continue
+
+    if col_type in NUMERIC_SCHEMA:
+      if transform not in NUMERIC_TRANSFORMS:
+        raise ValueError(
+            'Transform %s not supported by schema %s' % (transform, col_type))
+    elif col_type == STRING_SCHEMA:
+      if transform not in CATEGORICAL_TRANSFORMS + TEXT_TRANSFORMS:
+        raise ValueError(
+            'Transform %s not supported by schema %s' % (transform, col_type))
+    else:
+      raise ValueError('Unsupported schema type %s' % col_type)
+
+  if num_key_transforms != 1 or num_target_transforms != 1:
```

Comment:
As discussed key seems optional and we can allow multiple keys. I guess making it optional might involves more code changes but allowing multiple keys seems not a big change?